Implement rpmkeys --rebuild

ffesti opened the pull request:
This rebuilds the public key storage.

The keys from a previously used key store can be added --from KEYSTORE
Keys are stored in the latest format.

Resolves: #3347



Simplified implementation of #3474. Currently does not do key verification. That's something we might want to consider to do an key loading instead of the actual rebuild.

pmatilai commented on lib/rpmts.cc:
```diff
@@ -406,6 +406,50 @@ rpmRC rpmtsImportPubkey(const rpmts ts, const unsigned char * pkt, size_t pktlen
+    keystore_openpgp_cert_d ks_opengpg = {};
+    rpmKeyringIterator iter = NULL;
+
+    if (rpmtsOpenDB(txn->ts, (O_RDWR|O_CREAT))) {
```
This looks just wrong. If we're rebuilding an fs or openpgp keyring, the rpmdb doesn't come into play at all.

ffesti replied:
Fixed. There was quite a mess of part relying one the surrounding code opening the rpmdb and part in the keystore opening the rpmdb themselves. All code paths in the rpmdb key store should now open the rpmdb.

pmatilai commented on lib/rpmts.cc:
```diff
@@ -406,6 +406,50 @@ rpmRC rpmtsImportPubkey(const rpmts ts, const unsigned char * pkt, size_t pktlen
+    }
+    for (keystore *ks : std::vector<keystore*>
+	     {&ks_fs, &ks_rpmdb, &ks_opengpg}) {
+	ks->delete_store(txn);
```
We shouldn't be deleting anything we were not asked to. Only the source.

ffesti replied:
Done.

pmatilai reviewed (changes requested):
It's not a merge we want in --from mode, but move. Like with --rebuilddb. I thought that was clear in the ticket but I guess not. And like said above, we shouldn't be deleting anything we were not specifically asked to touch.

rpmKeyringSize() is not used here and shouldn't be added without any tests.

pmatilai commented on lib/keystore.hh:
```diff
@@ -10,19 +10,23 @@ namespace rpm {
+    virtual rpmRC delete_store(rpmtxn txn) = 0;
 
     virtual ~keystore() = default;
+    keystore(std::string n): name(n) {};
```
Something was nagging me about this name thing all along, and now I know what it was:  
We're now duplicating info with getKeystore() here. Instead, the name should be a static property, and then we should have an std::array of available types. And then getKeystore() can use that array to look up the right thing from the classes themselves. Like the rpmdb backend does. And, when you have a function that returns a backend based on a name, both getKeystore() and the --from logic can use that.

pmatilai replied:
Actually gcc gives a pretty helpful hint here when trying to add something like `static std::string name = "fs";` to the keystore:
> /home/pmatilai/repos/rpm/lib/keystore.hh:22:30: error: ‘constexpr’ needed for in-class initialization of static data member ‘const std::string rpm::keystore_fs::name’ of non-integral type [-fpermissive]
   22 |     static const std::string name = "fs";

`static constexpr std::string name = "fs";` merrily compiles.

pmatilai replied:
Anyway, the getKeystore() approach in the PR is totally okay - the knowledge is contained in exactly one place now.

pmatilai commented on tools/rpmkeys.cc:
```diff
@@ -149,7 +155,8 @@ int main(int argc, char *argv[])
 
-    if (args == NULL && mode != MODE_LISTKEY && mode != MODE_EXPORTKEY)
+    if (args == NULL && mode != MODE_LISTKEY && mode != MODE_EXPORTKEY &&
+	mode != MODE_REBUILD)
```
Indent nitpicks: this is confusingly on the same indent level as the codeblock below.

ffesti replied:
If I only could remember how to indent this properly...

pmatilai replied:
For crying out loud. It's not like you need to remember 21 digits of pi here. The issue is *ambiguity* due to identical indentation so to fix you just need to change how it ends up laid out, such as:

a) indent the condition continuation deeper.
b) put the opening brace on the next line (unlike the usual style), this provides the brain sufficient clue about the matter
c) avoid the whole issue by moving part of the condition elsewhere, such as using an aptly named helper variable

Quite often it's c) that makes for most readable code.

ffesti replied:
OK, so the reason is that there are too many, all not so great options. Went with a) already.

pmatilai replied:
Yeah a) and b) are mechanically applicable anywhere, c) requires thinking.

As a reminder, CONTRIBUTING.md has advice too:
> If in doubt, 'indent -kr' generally produces acceptable formatting.

ffesti commented:
> It's not a merge we want in --from mode, but move. Like with --rebuilddb. I thought that was clear in the ticket but I guess not. And like said above, we shouldn't be deleting anything we were not specifically asked to touch.

Sure. My test cases now run into automatically generated keys for signing builds which are for some reason in the previously empty keystore. Well, were in the keystore...

> rpmKeyringSize() is not used here and shouldn't be added without any tests.
Dropped.

pmatilai commented:
There's a conflict in the man page, preventing CI from running.

pmatilai commented:
> > Sure. My test cases now run into automatically generated keys for signing builds which are for some reason in the previously empty keystore. Well, were in the keystore...

This isn't verifying any signatures, so you can either blast the default keystore contents, or use a different keyring path for your tests. Or, filter the fingerprint out of the output, the name part is predictable.

pmatilai commented on lib/rpmts.cc:
```diff
@@ -406,6 +399,45 @@ rpmRC rpmtsImportPubkey(const rpmts ts, const unsigned char * pkt, size_t pktlen
+    rpmKeyring keyring = rpmtsGetKeyring(ts, 1);
+    keystore_fs ks_fs = {};
+    keystore_rpmdb ks_rpmdb = {};
+    keystore_openpgp_cert_d ks_opengpg = {};
```
All the three above variables are unused and should be removed.

The weird thing is, why isn't the compiler warning about these? :flushed:

ffesti commented:
> > > Sure. My test cases now run into automatically generated keys for signing builds which are for some reason in the previously empty keystore. Well, were in the keystore...
> 
> This isn't verifying any signatures, so you can either blast the default keystore contents, or use a different keyring path for your tests. Or, filter the fingerprint out of the output, the name part is predictable.

Well, I am not that worried about the test cases themselves. I am wondering whether nuking this key may create issues during transition from one keystore to another if the key is just created at the same time.

pmatilai commented:
> Well, I am not that worried about the test cases themselves. I am wondering whether nuking this key may create issues during transition from one keystore to another if the key is just created at the same time.

RPMTEST_CHECK() cases run serially within the group, and RPMTEST_SETUP_RW() groups each run in their own image.

ffesti commented:
Oh, so the test suite automatically creates these keys. This is not some automatic thing in RPM itself. So nevermind.

pmatilai commented:
Looks like we're approaching the finishing line now. Aside from the above remarks, we should have
- tests for rebuilding *without* --from for all backends
- a test for an invalid --from
- rpmdb keystore being the very special snowflake it is, it wouldn't hurt to test that an installed package in the db survives a rebuild
- there doesn't seem to be a test for --from=openpgp

These seem more important to me than testing for testing for those unrelated keystores in each case, but if you want to keep them that's ok with me.

pmatilai commented:
One more note on writing tests in general, I've said this before: whenver you feel the output needs a separator of some kind, that's a signal to use a new RPMTEST_CHECK(). Not a separator line. RPMTEST_CHECK() is easier for us to parse and understand, and ensures the exit codes are what they should be. That we have those separator lines in old testcases doesn't mean it's a good thing to do, those simply predate us realizing you can have multiple checks inside a test-group.

Not saying you have to rewrite all these tests to that style, but it would improve the tests.

ffesti commented:
OK, added the tests and dropped a few of the separators and split some tests that are just too long. For others I don't think the list, rebuild, list sequence does not benefit from stretching it out into multiple tests.

The other comments about the code should be adressed with the exception  of the keystore names. They are just dropped for now re-using the code from rpmts.cc for creating the keystore by name in the keystore module itself.

pmatilai commented:
Aaargh, sorry my bad. The comments I'm talking about are in "pending" state, I seem to have accidentally started a review when the intention was just to make individual comments.

pmatilai reviewed (changes requested):
Bunch of stuff from this morning's (and apparently some older too) review round. Intention was to have this all visible immediately, sorry!

pmatilai commented on docs/man/rpmkeys.8.scd:
```diff
@@ -57,6 +59,18 @@ Erase the key(s) designated by _FINGERPRINT_. For example:
+Recreate the public key storage. Update to the latest format and drop
+unreadable keys.
+
+Adding *--from* _BACKEND_ will use the keys from the given
```
Actually scratch that, this can be addressed along when the rest of the rpmkeys man page is reformatted to common style.

ffesti replied:
To late.

ffesti commented on tests/rpmsigdig.at:
```diff
@@ -2001,3 +2001,134 @@ rpm -qp --qf "[%{filenames}:%{filesignatures}\n]" hello-2.0-1.x86_64-badima.rpm
+echo ----------
+runroot rpmkeys --define "_keyring rpmdb" --list
+echo ----------
+runroot rpmkeys --define "_keyring openpgp" --list
```
The opposite is true. We destroy the openpgp backend here as it shares the same space as the fs backend. The test verifies that the key is gone now.

pmatilai replied:
Oh, right. I was somehow under the impression that fs and openpgp keystores used separate paths, but that's not the case.

pmatilai commented on lib/keystore.cc:
```diff
@@ -97,6 +114,13 @@ static rpmRC write_key_to_disk(rpmPubkey key, string & dir, string & filename, i
 
+static rpmRC delete_file_store(std::string path)
+{
+    fs::remove_all(path);
```
Oh, one final thing. Sorry for missing earlier, this is belongs to a whole new class of issues to watch out for:
This can throw an exception that aborts the program whereas we should just return a success/failure error code.
So this needs to either use the non-throwing version, or catch possible exceptions.

ffesti replied:
catching  `std::filesystem::filesystem_error` here is kinda straight forward. I wonder if we need a general policy on `std::bad_alloc`? We have the special alloc functions all over the code. I guess we want to keep that behaviour overall.

pmatilai replied:
Oh, the filesystem part is all I meant, sorry I guess I wasn't clear on that.

bad_alloc we should just let through everywhere - the rpm policy has always been to just die on alloc failures because continuing is going to be futile.

pmatilai replied:
So yeah just drop the bad_alloc check. We should document this someplace... okay, CONTRIBUTING has a section where this naturally falls into.

pmatilai replied:
stackoverflow agrees: https://stackoverflow.com/questions/9456728/how-to-deal-with-bad-alloc-in-c

> In general you cannot, and should not try, to respond to this error. bad_alloc indicates that a resource cannot be allocated because not enough memory is available. In most scenarios your program cannot hope to cope with that, and terminating soon is the only meaningful behaviour.

pmatilai replied:
There goes: https://github.com/rpm-software-management/rpm/commit/4c144d0825f975106e4ae1ae74c265b66391007e

pmatilai reviewed (approved):
:partying_face:
